src/models/history/tree-manager.ts

```diff
@@ -57,12 +77,137 @@ export const TreeManager = types
+    const validateCommentableDocument = 
+      getFirebaseFunction<ICommentableDocumentParams>("validateCommentableDocument_v1");
+
+    // We'll need provide this context to the tree manager either through a MST env or volatile prop
+
+    const {userContext, documentMetadata, firestore} = self;
+    if (!userContext || !documentMetadata || !firestore || !userContext.uid) {
+      console.warn("cannot record history entry because environment is not valid", 
+        {userContext, documentMetadata, firestore});
+      return;
+    }
+
+    // create a document if necessary
+    // TODO: need to figure how to access firestore
+
+    const networkDocKey = networkDocumentKey(userContext.uid, documentMetadata.key, userContext.network);
+    const documentPath = `documents/${networkDocKey}`;
+    new Promise<DocumentQueryType>((resolve, reject) => {
+      // It is crazy that we get this on every request
+      // This document should be provided as view, but we can't do async
+      // views. So I guess it is really an async action that fetchs the document and the view 
+      // will be undefined until then.
+      // so then we use something like mobx's `when()` to block the async action until the 
+      // view is updated.
+      const documentRef = firestore.doc(documentPath);
+      documentRef.get()
+        .then(docSnapshot => {
+          if (docSnapshot.exists) {
+            resolve(docSnapshot.data() as DocumentQueryType);          
+          } else {
+            resolve(
+              validateCommentableDocument({context: userContext, document: documentMetadata})
+              .then(result => result.data)
+            );
+          }
+        });
+      })
+    //   const unsubscribeDocListener = documentRef.onSnapshot({
+    //     next: docSnapshot => {
+    //       unsubscribeDocListener?.();
+    //       console.log("found existing document for key", documentMetadata.key);
+    //       resolve(docSnapshot.data() as DocumentQueryType);          
+    //     },
+    //     error: readError => {
+    //       unsubscribeDocListener?.();
+    //       // an error presumably means that the document doesn't exist yet, so we create it
+    //       console.log("creating document for key", documentMetadata.key);
+    //       resolve(
+    //         validateCommentableDocument({context: userContext, document: documentMetadata})
+    //         .then(result => result.data)
+    //       );
+    //     }
+    //   });
+    // })
+    .then(document => {
+      // add a new document for this history entry
+      const historyEntryPath = firestore.getFullPath(`${documentPath}/historyEntries`);
+
+      // FIXME-HISTORY: this should use the id of the history entry, that way a specific entry can
+      // be found again quickly. We plan to support the idea of providing links to teachers
+      // to open specific spots in the history. It seems better to use these ids than indexes,
+      // but maybe indexes are fine.
```

Comment:
I would think that ids would be preferable to indices.

Reply:
The reason I'm wishy-washy here has to do with document ordering. But having thought through that a bit, the name of the document in firestore doesn't really matter so we might as well use the id. Being able to lookup the history entry by this id doesn't really matter. We are going to have to use a query to get the document so as long as the id is stored in the document it doesn't really mater what the document's "name" is.